fxevent/event.go

```diff
@@ -30,30 +31,39 @@ type Event interface {
 }
 
 // Passing events by type to make Event hashable in the future.
-func (*LifecycleHookStart) event() {}
-func (*LifecycleHookStop) event()  {}
-func (*ProvideError) event()       {}
-func (*Supply) event()             {}
-func (*Provide) event()            {}
-func (*Invoke) event()             {}
-func (*InvokeError) event()        {}
-func (*StartError) event()         {}
-func (*StopSignal) event()         {}
-func (*StopError) event()          {}
-func (*Rollback) event()           {}
-func (*RollbackError) event()      {}
-func (*Running) event()            {}
-func (*CustomLoggerError) event()  {}
-func (*CustomLogger) event()       {}
-
-// LifecycleHookStart is emitted whenever an OnStart hook is executed
-type LifecycleHookStart struct {
+func (*LifecycleHookExecuting) event() {}
+func (*LifecycleHookExecuted) event()  {}
+func (*ProvideError) event()           {}
+func (*Supply) event()                 {}
+func (*Provide) event()                {}
+func (*Invoke) event()                 {}
+func (*InvokeError) event()            {}
+func (*StartError) event()             {}
+func (*StopSignal) event()             {}
+func (*StopError) event()              {}
+func (*Rollback) event()               {}
+func (*RollbackError) event()          {}
+func (*Running) event()                {}
+func (*CustomLoggerError) event()      {}
+func (*CustomLogger) event()           {}
+
+// LifecycleHookExecuting is emitted before an OnStart hook is about to be executed.
+type LifecycleHookExecuting struct {
+	// FunctionName is the name of the hook being executed.
+	FunctionName string
+	// CallerName is the name of the caller that appended the hook.
 	CallerName string
+	// Method is the lifecycle hook method getting called.
+	Method string
```

Comment:
Would it be overkill to introduce a LifecycleHookType enum here and use that with a Type field?

```
type LifecycleHookType int

const (
  LifecycleHookOnStart LifecycleHookType = iota + 1
  LifecycleHookOnStop
)

// We' can hand-write it, or use stringer or enumer to generate a String method
// for the enum. The trimprefix flag will be needed to make it be only "OnStop" or "OnStart".

type LifecycleHookExecuting struct {
  // ...
  Type LifecycleHookType
}
```

Reply:
I'll actually do this as a follow-up PR because it looks like we could use this enum in lifecycle implementation as well. Thanks for the suggestion!